csp_billing_adapter_local/plugin.py

```diff
@@ -111,3 +114,79 @@ def save_csp_config(
 ):
     """Save specified content as local storage csp_config contents."""
     update_csp_config(config, csp_config, replace=True)
+
+
+@csp_billing_adapter.hookimpl(trylast=True)
+def get_usage_data(config: Config):
+    """
+    Retrieves the current usage report from the application API
+
+    :param config: The application configuration dictionary
+    :return: Return a dict with the current usage report
+    """
+    usage_data = _make_request(config.get('api'))
+    usage_data_list_name = ''.join(list(usage_data.keys()))
+
+    if usage_data_list_name in config:
+        return _extract_usage(
+            usage_data[usage_data_list_name],
+            config[usage_data_list_name]
+        )
+    if 'usage_metrics' in usage_data_list_name:
+        raise CSPBillingAdapterException(
+            'Config missing usage metrics section'
+        )
+    raise CSPBillingAdapterException('Unrecognized API response')
+
+
+def _extract_usage(
+    api_usage_metrics: list,
+    config_usage_metrics: dict
+):
+    """Parse the response from API to the expected structure."""
+    usage_metrics = {}
+    missing_metrics = []
+    for usage_metric_info in api_usage_metrics:
+        usage_metric_name = usage_metric_info.get('usage_metric')
+        if usage_metric_name not in config_usage_metrics:
+            missing_metrics.append(usage_metric_name)
+        usage_metrics[usage_metric_name] = usage_metric_info.get('count')
```

Comment:
We probably want `0` as default return value here instead of `None`

Reply:
was not sure what default we wanted there, 
when I picked that I did not put `0` because wouldn't `0` be misleading ? 
Meaning the usage count has been `0` , when the reality is that that count is not present ?

Reply:
"there has been no usage" vs "we dont know what the usage is"

Reply:
maybe `-1` to indicate we do not know the usage ?
`usage_metric_info.get('count', -1)`

Reply:
Non-negative integer (0). We don't care if it's an implicit or explicit zero.

Reply:
does that mean we do not care between "no usage count" vs "we dont know what the usage count is" ?

Reply:
The condition of "no usage count" is not allowed to exist, the application is expected to report. S AS such we have 2 choices, we either complain, i.e. in the current implementation we'd add
```
if not usage_metrics[usage_metric_name]:
    raise THE_APPLICATION_BROKE_ITS_PROMISE
```
or we decide that if he application reports no `count` entry in the structure it is equivalent to having no usage, i.e. `usage_metric_info.get('count', 0)`. We cannot stick `None` into the structure that trickles through to the reporting code. Using 0, i.e. ignoring that the application breaks it's promise is safe as we are not inflicting any charges on the customer.

Reply:
> "no usage count"

I meant `count` could be 0, I understand that is accepted

`usage_metric_info.get('count', 0)` is as it is now, 
I see no problem from the customer perspective 
but wouldnt we want to catch if the application 
API is behaving strangely (e.g. missing `count`) ?

In an scenario where there has been usage,
meaning customer should be charged and the 
application API response does no contain `count`,
no charge would be applied

Reply:
Seems better to not trigger an exception scenario. But it could be logged at a warning or info level.